tool/dart_model_generator/lib/generate_dart_model.dart

```diff
@@ -0,0 +1,136 @@
+    final schema = JsonSchema.create(
+        File('schemas/dart_model.schema.json').readAsStringSync());
+    for (final def in schema.defs.entries) {
+      result.add(_generateExtensionType(def.key, def.value));
+    }
+
+    File('pkgs/dart_model/lib/src/dart_model.g.dart').writeAsStringSync(
+        DartFormatter().formatSource(SourceCode(result.join('\n'))).text);
+  }
+
+  String _generateExtensionType(String name, JsonSchema definition) {
+    final result = StringBuffer();
+
+    final jsonType = switch (definition.type) {
+      SchemaType.object => 'Map<String, Object?> node',
+      SchemaType.string => 'String string',
+      _ => throw UnsupportedError('Schema type ${definition.type}.'),
+    };
+    result.writeln('extension type $name.fromJson($jsonType) {');
+
+    final propertyMetadatas = definition.properties.entries
+        .map((e) => _readPropertyMetadata(e.key, e.value))
+        .toList();
+    if (definition.type == SchemaType.object) {
+      result.writeln('  $name({');
+      for (final property in propertyMetadatas) {
+        result.writeln(switch (property.type) {
+          PropertyType.object =>
+            '${property.elementTypeName}? ${property.name},',
+          PropertyType.bool => 'bool? ${property.name},',
+          PropertyType.string => 'string? ${property.name},',
+          PropertyType.list =>
+            'List<${property.elementTypeName}>? ${property.name},',
+          PropertyType.map =>
+            'Map<String, ${property.elementTypeName}>? ${property.name},',
+        });
+      }
+      result.writeln('}) : this.fromJson({');
+      for (final property in propertyMetadatas) {
+        result.writeln('if (${property.name} != null) '
+            "'${property.name}': ${property.name},");
+      }
+      result.writeln('});');
+    } else if (definition.type == SchemaType.string) {
+      result.writeln('$name(String string) : this.fromJson(string);');
+    }
+
+    for (final property in propertyMetadatas) {
+      result.writeln(switch (property.type) {
+        PropertyType.object =>
+          // TODO(davidmorgan): use the extension type constructor instead of
+          // casting.
+          '${property.elementTypeName} get ${property.name} => '
+              'node[\'${property.name}\']!'
+              ' as ${property.elementTypeName};',
+        PropertyType.bool => 'bool get ${property.name} => '
+            'node[\'${property.name}\']! as bool;',
+        PropertyType.string => 'String get ${property.name} => '
+            'node[\'${property.name}\']! as String;',
+        PropertyType.list =>
+          'List<${property.elementTypeName}> get ${property.name} => '
+              'node[\'${property.name}\']! '
+              'as List<${property.elementTypeName}>;',
+        PropertyType.map =>
+          'Map<String, ${property.elementTypeName}> get ${property.name} => '
+              'node[\'${property.name}\']! '
+              'as Map<String, ${property.elementTypeName}>;',
```

Comment:
Same issue here with the value types, pretty sure this cast will fail

Reply:
Yup yup, added test coverage and changed to use "cast". Thanks.